design/httpproxy-conflict.md

```diff
@@ -0,0 +1,82 @@
+# HTTPProxy Conflict Resolution Design
+
+Status: Draft
+
+## Abstract
+
+Configuration errors in HTTPProxy resources can result in objects being marked as invalid and conversely will not serve traffic.
+These errors can come from invalid route configuration, missing service references, or invalid includes to other resources. 
+
+## Background
+HTTPProxy is designed as a multi-team aware Ingress, reduce annotation sprawl by providing a sensible home for configuration items.
+Additionally, HTTPProxy offers a feedback loop to users via a `Status` field on the object.
+Contour has further extended this to add `Conditions` which allows for a set of `Errors` and `Warnings` which inform users of errors in the HTTPProxy resource.
+
+When the first error or warning is detected, these fields are updated and in the status of the HTTPProxy object, however if there are more errors, they are not all shown to the user, only the first.
+There are ongoing efforts to change this logic to inform of all the errors or warnings dealing with a specific object.
+
+Contour currently invalidates the entire resource if a configuration error is encountered which can cause downtime for specific routes. 
+
+## Goals
+- Serve valid traffic configurations if portions of the spec are invalid
+
+## Non Goals
+- Surface a complete set errors or warnings for a resource (this is a different issue)
+
+## High-Level Design
+Contour will set the error or warning condition when an error is encountered, but do its best to still serve valid configurations.
+
+- If an invalid route is encountered, Contour will return a `404` HTTP status code to the requester
+- If an invalid service is encountered,Contour will return a`503` HTTP status code to the requester
+
+It's difficult to find errors and block them from breaking valid configurations due to how Contour currently processes objects.
+Contour takes a passive look to the cluster by reacting to changes after they have changed. 
+Due to this, Contour cannot block changes to resources before they are committed like an `Admission Controller` would be able to do (See alternatives: `HTTPProxyInstance`)
+
+## Detailed Design
```

Comment:
Should the usability aspect be discussed here, I've had following concerns in my mind:

* When the `HTTPProxy` handles traffic for the (valid) part of the configuration, the configuration errors that were previously very visible will become warnings and might be more difficult to notice.
* Previously the user got detailed error message (though only for the first error). After turning invalid routes/services to warnings, the final error message will be something general like "all routes are invalid". The details must be looked from warnings.  This might seem especially strange for the user when there was just single route in the `HTTPProxy`.

Reply:
I agree that the usability should be discussed here - I think that we'll need to make it easier for people to get all the conditions for an object somehow. Some experiments I did earlier didn't show any easy way to summarise a list of things into a field using the jsonPath tools that the CRD markup allows, but I'd love to be proved wrong on this one.

Reply:
We can try and summarize the bits into a single line, but I thought the idea was folks would need to now dig into the object to figure out what's wrong, or we build some sort of plugin/octant plugin, etc to assist.